src/chain/split.rs

```diff
@@ -0,0 +1,939 @@
+    fn from_specific(specific: Self::SpecificKey) -> Self;
+}
+
+/// This enum allows users to key into splittable list-like structures based on
+/// the sequence in which an item appears in the list. It also has an option for
+/// keying into any items that were left over in the sequence.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum ListSplitKey {
+    /// Key into an item at a specific point in the sequence.
+    Sequential(usize),
+    /// Key into any remaining items that were not covered by the sequential keys.
+    Remaining,
+}
+
+impl FromSequential for ListSplitKey {
+    fn from_sequential(seq: usize) -> Self {
+        ListSplitKey::Sequential(seq)
+    }
+}
+
+impl ForRemaining for ListSplitKey {
+    fn for_remaining() -> Self {
+        ListSplitKey::Remaining
+    }
+}
+
+/// This is a newtype that implements [`Splittable`] for anything that can be
+/// turned into an iterator, but always splits it as if the iterator is list-like.
+///
+/// If used with a map-type (e.g. [`HashMap`] or [BTreeMap]), the items will be
+/// the `(key, value)` pairs of the maps, and the key for the split will be the
+/// order in which the map is iterated through. For [`BTreeMap`] this will always
+/// be a sorted order (e.g. alphabetical or numerical) but for [`HashMap`] this
+/// order can be completely arbitrary and may be unstable.
+pub struct SplitAsList<T: 'static + Send + Sync + IntoIterator> {
+    pub contents: T,
+}
+
+impl<T: 'static + Send + Sync + IntoIterator> SplitAsList<T> {
+    pub fn new(contents: T) -> Self {
+        Self { contents }
+    }
+}
+
+impl<T> Splittable for SplitAsList<T>
+where
+    T: 'static + Send + Sync + IntoIterator,
+    T::Item: 'static + Send + Sync,
+{
+    type Key = ListSplitKey;
+    type Item = (usize, T::Item);
+
+    fn validate(_: &Self::Key) -> bool {
+        // We don't know if there are any restrictions for the iterable, so just
+        // return say all keys are valid
+        true
+    }
+
+    fn next(key: &Option<Self::Key>) -> Option<Self::Key> {
+        if let Some(key) = key {
+            match key {
+                ListSplitKey::Sequential(k) => Some(ListSplitKey::Sequential(*k + 1)),
+                ListSplitKey::Remaining => None,
+            }
+        } else {
+            Some(ListSplitKey::Sequential(0))
+        }
+    }
+
+    fn split(self, mut dispatcher: SplitDispatcher<'_, Self::Key, Self::Item>) -> OperationResult {
+        for (index, value) in self.contents.into_iter().enumerate() {
+            match dispatcher.outputs_for(&ListSplitKey::Sequential(index)) {
+                Some(outputs) => {
+                    outputs.push((index, value));
+                }
+                None => {
+                    if let Some(outputs) = dispatcher.outputs_for(&ListSplitKey::Remaining) {
+                        outputs.push((index, value));
+                    }
+                }
+            }
+        }
```

Comment:
iiuc we iterate through everything even if some of the indexes are not used, since `SplitDispatcher` keeps track of the connections, is it possible to use that to avoid iterating?